### PR TITLE
Issue 4231: Use system truststore for TLS when client truststore is not specified

### LINKED ISSUE #4231: ClientConfig fails to use system truststore if client truststore is not specified
**Problem description**
ClientConfig [API documentation](https://github.com/pravega/pravega/blob/master/client/src/main/java/io/pravega/client/ClientConfig.java#L60) says that trustStore is optional. If it is null or empty, the system trust store is used.

However, a user encountered the following error in the client when connecting to a TLS-enabled server. The server was using certificates signed by Let's Encrypt CA. Let's Encrypt CA's certificate was in the client's Java system truststore. 
```
javax.net.ssl.SSLException: unable to setup trustmanager
```
Specifying client truststore should not be necessary if certificates issued by the system's trusted CAs are used by the endpoints.

**Problem location**
https://github.com/pravega/pravega/blob/1cedca8d3b36782139ad7827ee89682de24f9689/client/src/main/java/io/pravega/client/netty/impl/ConnectionPoolImpl.java#L260-L282

**Suggestions for an improvement**

### PR DESCRIPTION
**Pull request by ravisharda**
**Change log description**  
Enable use of system truststore for client-server TLS communications when client truststore is unspecified.

**Purpose of the change**  
Fixes #4231. 

**What the code does**  
Here's what the code does: 
* Removes the use of `FingerprintTrustManagerFactory` when client truststore is null or empty. 
* Instead, sets Netty `SslContextBuilder trustManager` as null, so that system truststore is used 

**How to verify it**  
All unit and integration tests should pass. A couple of focused tests used for testing the PR are as follows: 
1. Verify that clients are able to read/write events over TLS-enabled channels when client truststore is specified. 
   Run tests in `TlsEnabledInProcPravegaClusterTest` for the purpose. 
2. Verify that client are able to read/write events when the CA certificate is not specified via `ClientConfig` but is present in the system truststore. 
   1. Create new TLS material for Pravega standalone using steps described [here](https://github.com/pravega/pravega/wiki/Creating-TLS-material-for-Pravega-security). 
   2. Modify Pravega standalone config to enable TLS and Auth. Point TLS params to newly created material. Here's how my configuration looked like: 
      ```
      singlenode.enableAuth=true
      singlenode.userName=admin
      singlenode.passwd=1111_aaaa
      singlenode.passwdFile=../config/passwd

      singlenode.enableTls=true
      singlenode.keyFile=/home/xyz/temp/pki-client-truststore/server-key.key
      singlenode.certFile=/home/xyz/temp/pki-client-truststore/server-cert.crt
      singlenode.keyStoreJKS=/home/xyz/temp/pki-client-truststore/server.keystore.jks
      singlenode.keyStoreJKSPasswordFile=/home/xyz/temp/pki-client-truststore/server.keystore.jks.passwd
      ```
   3.  Run a reader/writer client against the standalone server (like the one shown [here](https://github.com/pravega/pravega-samples/tree/master/pravega-client-examples/src/main/java/io/pravega/example/secure), without specifying the client truststore. You should see an error like the following on the client-side: 
 ```
       2019-09-30 23:13:19.467 [grpc-default-executor-1] WARN  i.p.c.stream.impl.ControllerImpl.warn(285) - gRPC call for createScope with trace id 3903548272740574148 failed with server error.
io.grpc.StatusRuntimeException: UNAVAILABLE: io exception
	at io.grpc.Status.asRuntimeException(Status.java:530)
at io.grpc.stub.ClientCalls$StreamObs
      ....
     Caused by: javax.net.ssl.SSLHandshakeException: General OpenSslEngine problem
	at io.netty.handler.ssl.ReferenceCountedOpenSslContext$AbstractCertificateVerifier.verify(ReferenceCountedOpenSslContext.java:629)
	at io.netty.internal.tcnative.SSL.readFromSSL(Native Method)
	at io.netty.handler.ssl.ReferenceCountedOpenSslEngine.readPlaintextData(ReferenceCountedOpenSslEngine.java:511)
...
Caused by: sun.security.validator.ValidatorException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
	at sun.security.validator.PKIXValidator.doBuild(PKIXValidator.java:397)
	at sun.security.validator.PKIXValidator.engineValidate(PKIXValidator.java:302)
	at sun.security.validator.Validator.validate(Validator.java:262)
	at sun.security.ssl.X509TrustManagerImpl.validate(X509TrustManagerImpl.java:330)
at sun.sec       

```
   4. Now, add the CA certificate to the system truststore: 
       ```
        $ sudo keytool -import -trustcacerts -keystore /usr/lib/jvm/java-8-openjdk-amd64/jre/lib/security/cacerts -storepass changeit -alias newlocalhost -import -file ca-cert.crt
       ```
  5. Now, run the reader/writer test again. This time the tests should succeed.

### CODE REVIEW

File: client/src/main/java/io/pravega/client/netty/impl/ConnectionPoolImpl.java [hunk 1]
```diff
@@ -263,16 +261,18 @@ private SslContext getSslContext() {
...
             try {
                 SslContextBuilder clientSslCtxBuilder = SslContextBuilder.forClient();
+
                 if (Strings.isNullOrEmpty(clientConfig.getTrustStore())) {
-                    clientSslCtxBuilder = clientSslCtxBuilder.trustManager(FingerprintTrustManagerFactory
-                                                                                   .getInstance(FingerprintTrustManagerFactory.getDefaultAlgorithm()));
-                    log.debug("SslContextBuilder was set to an instance of {}", FingerprintTrustManagerFactory.class);
+                    log.debug("Client truststore wasn't specified.");
+                    File clientTruststore = null; // variable for disambiguating method call
+                    clientSslCtxBuilder.trustManager(clientTruststore);
```
Comment (shrids) [inline client/src/main/java/io/pravega/client/netty/impl/ConnectionPoolImpl.java]: ok, setting trustManager(null) uses the system default trust store

**Comment by ravisharda** [comment]
> This can be merged once @ravisharda completes verifying the fix.

> Needs more testing before it can be merged.

The testing is now complete. The tests cases used are described in the issue description. This PR may be merged, after approval.

File: client/src/main/java/io/pravega/client/netty/impl/ConnectionPoolImpl.java [hunk 2]
```diff
@@ -257,22 +256,25 @@ public void initChannel(SocketChannel ch) throws Exception {
...
```
Comment (gaddamas) [inline client/src/main/java/io/pravega/client/netty/impl/ConnectionPoolImpl.java]: This block of setting the null for trustManager is not necessary, but no harm If you want to make it explicit.

Comment (ravisharda) [reply client/src/main/java/io/pravega/client/netty/impl/ConnectionPoolImpl.java]: Yes, I wanted to make it explicit.

**Comment by EronWright** [comment]
Thanks this looks good.

**Comment by fpj** [comment]
The build for this PR is failing consistently. Are you sure this change is not inducing new test failures or causing an intermittent test to fail more often?

**Comment by ravisharda** [comment]
> The build for this PR is failing consistently. Are you sure this change is not inducing new test failures or causing an intermittent test failure to occur more often?

@fpj all the tests that have failed so far are unrelated to the PR, and I've been noticing some of these tests failing occasionally in other PR builds too. I believe these tests are flaky. I've triggered a rebuild yet again a few minutes back. Please also note that my changes up to my last commit `b780e97` built successfully. The builds that failed later were the merges back from the master to this branch.

File: client/src/main/java/io/pravega/client/netty/impl/ConnectionPoolImpl.java (hunk 1 above)
Comment (andreipaduroiu) [reply client/src/main/java/io/pravega/client/netty/impl/ConnectionPoolImpl.java]: You can also do `(File)null` so you don't have to declare a new variable just for this.

**Comment by fpj** [comment]
This test case just failed:

```
io.pravega.local.SecurePravegaClusterTest > testWriteAndReadEventWithValidClientConfig FAILED
    org.junit.runners.model.TestTimedOutException: test timed out after 50000 milliseconds
```

The number of test failures and this latest test failure makes me skeptical about these changes.

**Comment by ravisharda** [comment]
> This test case just failed:
> 
> ```
> io.pravega.local.SecurePravegaClusterTest > testWriteAndReadEventWithValidClientConfig FAILED
>     org.junit.runners.model.TestTimedOutException: test timed out after 50000 milliseconds
> ```
> 
> The number of test failures and this latest test failure makes me skeptical about these changes.

The build has now passed (without any changes from my side). I had merged master into this branch of my fork and retriggered the build.

**Comment by maddisondavid** [comment]
Is there something that's delaying this PR being merged?  This needs to be merged in order to update the Flink Connectors to work with Pravega TLS.

**Pull request merged**